apis/core/v1alpha2/featuregate_webhook.go

```diff
@@ -0,0 +1,270 @@
+	}
+
+	ctx := context.Background()
+	var allErrors field.ErrorList
+
+	allErrors = append(allErrors, r.validateFeatureExistence(ctx, c)...)
+	allErrors = append(allErrors, r.validateFeatureConflicts(ctx, c)...)
+	allErrors = append(allErrors, r.validateStabilityPolicyOverride(oldObj)...)
+	allErrors = append(allErrors, r.validateFeatureForStabilityPolicyViolation(ctx, c)...)
+
+	if len(allErrors) == 0 {
+		return nil
+	}
+	return apierrors.NewInvalid(GroupVersion.WithKind("FeatureGate").GroupKind(), r.Name, allErrors)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *FeatureGate) ValidateDelete() error {
+	featuregatelog.Info("validate delete", "name", r.Name)
+	return nil
+}
+
+// validateFeatureForStabilityPolicyViolation validates the feature for any stability policy violation
+func (r *FeatureGate) validateFeatureForStabilityPolicyViolation(ctx context.Context, c client.Client) field.ErrorList {
+	var allErrors field.ErrorList
+
+	features := &FeatureList{}
+	if err := c.List(ctx, features); err != nil {
+		allErrors = append(allErrors, field.InternalError(field.NewPath("spec").Child("features"), err))
+		return allErrors
+	}
+
+	invalidFeatures := sets.String{}
+	for _, featureRef := range r.Spec.Features {
+		stabilityLevel, found := getFeatureStabilityLevel(features, featureRef.Name)
+		if !found {
+			// Feature doesn't exist and is validated in validateFeatureExistence method
+			continue
+		}
+		policy := GetPolicyForStabilityLevel(stabilityLevel)
+		// immutability validation
+		if policy.Immutable && !featureRef.SkipStabilityPolicyValidation && policy.DefaultActivation != featureRef.Activate {
+			invalidFeatures.Insert(featureRef.Name)
+		} else if policy.VoidsWarranty && !featureRef.SkipStabilityPolicyValidation { // warranty validation
+			invalidFeatures.Insert(featureRef.Name)
+		}
+	}
+
+	if len(invalidFeatures) != 0 {
+		allErrors = append(allErrors, field.Invalid(field.NewPath("spec").Child("features"), r.Spec.Features,
+			fmt.Sprintf("cannot toggle features due to policy violation: %v."+
+				"Set feature.skipStabilityPolicyValidation to true to override policy."+
+				"Once set to true, cannot be set back to false.", invalidFeatures.List())))
+	}
+	return allErrors
+}
+
+// validateStabilityPolicyOverride validates stability policy override
+func (r *FeatureGate) validateStabilityPolicyOverride(oldObject *FeatureGate) field.ErrorList {
+	var allErrors field.ErrorList
+
+	invalidFeatures := sets.String{}
+	for _, featureRef := range r.Spec.Features {
+		skipStabilityPolicyValidationFlagFromOldObj, found := getSkipStabilityPolicyValidationFieldForFeature(oldObject, featureRef.Name)
+		if !found {
+			// Feature doesn't exist in old object and need not be validated
+			continue
+		}
+		if !featureRef.SkipStabilityPolicyValidation && skipStabilityPolicyValidationFlagFromOldObj {
+			invalidFeatures.Insert(featureRef.Name)
+		}
+	}
+	if len(invalidFeatures) != 0 {
```

Comment:
I know it's impossible for length here to be negative, but it would be better to be more explicit about your intention.
```suggestion
	if len(invalidFeatures) > 0 {
```

Reply:
i think it shouldn't matter, but will update it